src/preact/component/intersection-observer-resize.js

```diff
@@ -0,0 +1,50 @@
+import {toWin} from '#core/window';
+
+import {useEffect} from '#preact';
+
+const ioForWindow = new WeakMap();
+const callbackMap = new WeakMap();
+
+/**
+ * Uses a shared IntersectionObserver per window instance to observe the given `ref`.
+ *
+ * @param {{current: Element}} ref
+ * @param {function(IntersectionObserverEntry)} callback
+ * @param {?Window} targetWin
+ */
+export function useIntersectionObserver(ref, callback, targetWin = null) {
+  useEffect(() => {
+    const node = ref.current;
+    if (!node) {
+      return;
+    }
+    // If a window instance is not provided, uses the one local to the given `ref`.
+    const win = targetWin ?? toWin(node.ownerDocument?.defaultView);
+    if (!win) {
+      return;
+    }
+    callbackMap.set(node, callback);
```

Comment:
Are you sure that no two hooks will ever track the same element? The way this is written there can only be a single callback for each node.

If you want to handle multiple, we'll need a  [Multimap](https://en.wikipedia.org/wiki/Multimap), aka `WeakMap<!Element, Array<function()>`.

A good example of handling multiple callback function is the `size-observer` that Dima recently wrote: https://github.com/ampproject/amphtml/blob/main/src/core/dom/layout/size-observer.js.


In fact, if we added multiple callback support to our [viewport-observer](https://github.com/ampproject/amphtml/blob/main/src/core/dom/layout/size-observer.js) I bet we could reuse it for this hook.

Reply:
Good point. I've updated `viewport-observer.js` functions so we can have multiple callbacks for each node and used that in this hook. PTAL.